pkg/istio-agent/health/health_probers_test.go

```diff
@@ -0,0 +1,174 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package health
+
+import (
+	"errors"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strconv"
+	"testing"
+	"time"
+
+	"istio.io/api/networking/v1alpha3"
+)
+
+func TestHttpProber(t *testing.T) {
+	tests := []struct {
+		desc                string
+		statusCode          int
+		expectedProbeResult ProbeResult
+		expectedError       error
+	}{
+		{
+			desc:                "Healthy - 200 status code",
+			statusCode:          200,
+			expectedProbeResult: Healthy,
+			expectedError:       nil,
+		},
+		{
+			desc:                "Unhealthy - 500 status code",
+			statusCode:          500,
+			expectedProbeResult: Unhealthy,
+			expectedError:       errors.New("status code was not from [200,400)"),
+		},
+		{
+			desc:                "Unhealthy - Could not connect to server",
+			statusCode:          -1,
+			expectedProbeResult: Unhealthy,
+			expectedError:       errors.New("get \"http://127.0.0.1:61023/test/health/check\": dial tcp 127.0.0.1:61023: connect: connection refused"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			server, port := createHTTPServer(tt.statusCode)
+			httpProber := HTTPProber{
+				Config: &v1alpha3.HTTPHealthCheckConfig{
+					Path:   "/test/health/check",
+					Port:   port,
+					Host:   "127.0.0.1",
+					Scheme: "http",
+				},
+			}
+
+			if tt.statusCode == -1 {
+				server.Close()
+			}
+
+			got, err := httpProber.Probe(time.Second)
+			if got != tt.expectedProbeResult || (err == nil && tt.expectedError != nil) || (err != nil && tt.expectedError == nil) {
+				server.Close()
+				t.Errorf("%s: got: %v, expected: %v, got error: %v, expected error %v", tt.desc, got, tt.expectedProbeResult, err, tt.expectedError)
+			}
+
+			server.Close()
+		})
+	}
+}
+
+func TestTcpProber(t *testing.T) {
+	tests := []struct {
+		desc                string
+		expectedProbeResult ProbeResult
+		expectedError       error
+	}{
+		{
+			desc:                "Healthy",
+			expectedProbeResult: Healthy,
+			expectedError:       nil,
+		},
+		{
+			desc:                "Unhealthy",
+			expectedProbeResult: Unhealthy,
+			expectedError:       errors.New("dial tcp 127.0.0.1:61023: connect: connection refused"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			server, port := createHTTPServer(200)
+			tcpProber := TCPProber{
+				Config: &v1alpha3.TCPHealthCheckConfig{
+					Host: "127.0.0.1",
+					Port: port,
+				},
+			}
+
+			if tt.expectedProbeResult == Unhealthy {
+				server.Close()
+			}
+
+			got, err := tcpProber.Probe(time.Second)
+			if got != tt.expectedProbeResult || (err == nil && tt.expectedError != nil) || (err != nil && tt.expectedError == nil) {
+				server.Close()
```

Comment:
`defer server.Close()`

Reply:
Since I create the server in the for loop, wouldn't it make more sense to close the server after each iteration? `defer` would defer closing the server until after the function is complete right?

Reply:
never mind, it would be deferring within the anonymous function scope. good catch 👍